### PR TITLE
Resolve incorrect uses of nullable values

### PR DESCRIPTION
**Pull request by dlkinney**
1. Enabled "Incorrect Uses of Nullable Values" compile warning (`CLANG_WARN_NULLABLE_TO_NONNULL_CONVERSION`) 
2. Made nonnull conversion explicit using a type cast where values are checked
3. Resolved remaining warnings about invalid use of nullable values
    1. Checked types where possible
    2. Updated `__nullable` and `__nonnull` return types and property declarations where appropriate -- some were incorrectly declared. For example, one property had a comment that a value would be `nil` under certain circumstances, but the property was `nonnull`. 
4. Resolved invalid use of nullable values in tests
5. Resolved invalid use of nullable values in the command line targets

### CODE REVIEW

File: Code/HTMLOrderedDictionary.h [hunk 1]
```diff
@@ -28,10 +28,10 @@ NS_ASSUME_NONNULL_BEGIN
 - (ObjectType)objectAtIndexedSubscript:(NSUInteger)index;
 
 /// Returns the key at index 0 in the dictionary, or nil if the dictionary is empty.
-@property (readonly, nonatomic) KeyType firstKey;
+@property (readonly, nonatomic, nullable) KeyType firstKey;
```
Comment (nolanw) [inline Code/HTMLOrderedDictionary.h]: This won't compile where nullability annotations are unavailable (see the CI build failure). We need to find some way to conditionally add `, nullable` only when it's available.

File: Code/HTMLOrderedDictionary.h [hunk 2]
```diff
@@ -28,10 +28,10 @@ NS_ASSUME_NONNULL_BEGIN
...
 
 /// Returns the key at index (count - 1) in the dictionary, or nil if the dictionary is empty.
-@property (readonly, nonatomic) KeyType lastKey;
+@property (readonly, nonatomic, nullable) KeyType lastKey;
```
Comment (nolanw) [inline Code/HTMLOrderedDictionary.h]: Ditto.

**Comment by nolanw** [comment]
This looks good! I've made a couple comments above, if you could kindly take a look then I'd love to merge this in.

`HTMLSupport.h` `#define`s fallbacks for nullability, but if we use them (i.e. use the underscored versions like `__nullable`) in the property declarations we'll have a dangling comma. Assuming the dangling comma doesn't compile, I think a lil macro that expands to `, nullable` iff nullability is available would be fine, and we can declare it just in `HTMLOrderedDictionary.h` as that looks to be the only affected header file.

Lemme know if any of that doesn't make sense!

**Comment by dlkinney** [comment]
Nolan,

It was trivial to move the `nullable` out of the property attributes and use `__nullable` after the type declaration of the property. This works properly for Xcode 6 and Xcode 7 (I think… waiting on Travis CI to agree).

**Comment by dlkinney** [comment]
Yay — TravisCI likes it. 

Please let me know if there is anything further you would like me to amend in the PR before you accept it.

**Comment by nolanw** [comment]
Oh, yep, great idea!

**Comment by nolanw** [comment]
Thanks so much!

**Pull request closed without merging**